cpp/diffraction/StaticStructureFactorDirect.cc

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) 2010-2020 The Regents of the University of Michigan
+// This file is from the freud project, released under the BSD 3-Clause License.
+
+#include <algorithm>
+#include <cmath>
+#include <complex>
+#include <limits>
+#include <random>
+#include <stdexcept>
+#include <tbb/concurrent_vector.h>
+
+#include "Eigen/Eigen/Dense"
+
+#include "Box.h"
+#include "ManagedArray.h"
+#include "NeighborQuery.h"
+#include "StaticStructureFactorDirect.h"
+#include "utils.h"
+
+/*! \file StaticStructureFactorDirect.cc
+    \brief Routines for computing static structure factors.
+*/
+
+namespace freud { namespace diffraction {
+
+StaticStructureFactorDirect::StaticStructureFactorDirect(unsigned int bins, float k_max, float k_min, unsigned int max_k_points)
+: m_max_k_points(max_k_points)
+{
+    if (bins == 0)
+    {
+        throw std::invalid_argument("StaticStructureFactorDirect requires a nonzero number of bins.");
+    }
+    if (k_max <= 0)
+    {
+        throw std::invalid_argument("StaticStructureFactorDirect requires k_max to be positive.");
+    }
+    if (k_min < 0)
+    {
+        throw std::invalid_argument("StaticStructureFactorDirect requires k_min to be non-negative.");
+    }
+    if (k_max <= k_min)
+    {
+        throw std::invalid_argument(
+            "StaticStructureFactorDirect requires that k_max must be greater than k_min.");
+    }
+
+    // We must construct two separate histograms, one for the counts and one
+    // for the actual correlation function. The counts are used to normalize
+    // the correlation function.
+    const auto axes = StructureFactorHistogram::Axes {std::make_shared<util::RegularAxis>(bins, k_min, k_max)};
+    m_structure_factor = StructureFactorHistogram(axes);
+    m_local_structure_factor = StructureFactorHistogram::ThreadLocalHistogram(m_structure_factor);
+    m_histogram = KBinHistogram(axes);
+    m_local_histograms = KBinHistogram::ThreadLocalHistogram(m_histogram);
+}
+
+void StaticStructureFactorDirect::accumulate(const freud::locality::NeighborQuery* neighbor_query,
+                                             const vec3<float>* query_points, unsigned int n_query_points,
+                                             unsigned int n_total, bool reuse_box)
+{
+    // Compute k vectors by sampling reciprocal space
+    const auto& box = neighbor_query->getBox();
+    if (box.is2D()) {
+        throw std::invalid_argument("2D boxes are not currently supported.");
+    }
+    const auto k_bin_edges = m_structure_factor.getBinEdges()[0];
+    const auto k_min = k_bin_edges.front();
+    const auto k_max = k_bin_edges.back();
+    if (!reuse_box || !m_k_grid_assigned) {
```

Comment:
@DomFijan I'm not sure if the `reuse_box` flag is consistent with the rest of freud. An alternative would be to resample if (`reset=True` or if the current box object is not equal to the previous box object). That is, we would store `m_box` as a member variable for the last used box object (many other classes do this), and re-sample if `reset || (m_box != box)`.

Reply:
Saving and comparing the box might be a good solution. It is also probably a more elegant then current solution.

Reply:
@DomFijan If you have time to implement that, go ahead. Otherwise I'll probably be able to look at it next weekend.

Reply:
I'll give this a go this week. I briefly check the == operator on the box class and it seems it only checks the lengths and tilts, but not PBCs. Do you think PBCs should also be checked in this case?

Reply:
@DomFijan Yes. PBCs are not well supported in freud.